### PR TITLE
Feature/issue 20

### PR DESCRIPTION
**Pull request by edouardmichelin**

### CODE REVIEW

File: app/src/main/java/com/studhub/app/domain/usecase/listing/GetListings.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import com.studhub.app.domain.repository.ListingRepository
+import kotlinx.coroutines.flow.Flow
+
+
+/**
+ * Use case for getting all listings from a given [repository]
+ *
+ * @param [repository] the repository which the use case will act on
+ */
+class GetListings constructor(private val repository: ListingRepository) {
```
Comment (raynasr) [inline app/src/main/java/com/studhub/app/domain/usecase/listing/GetListings.kt:14]: 'constructor' unnecessary

File: app/src/main/java/com/studhub/app/domain/usecase/listing/RemoveListing.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        return repository.removeListing(listing.id)
+    }
+
+
+    /**
+     * Removes the listing matching the given [listingId] from the [repository]
+     *
+     * @param [listingId] the ID of the listing to remove
+     */
+    suspend operator fun invoke(listingId: Long): Flow<ApiResponse<Boolean>> {
```
Comment (raynasr) [inline app/src/main/java/com/studhub/app/domain/usecase/listing/RemoveListing.kt:30]: rename to 'removeListingById' to differentiate from first invoke method which takes a Listing object

Comment (edouardmichelin) [reply app/src/main/java/com/studhub/app/domain/usecase/listing/RemoveListing.kt:30]: Here I would rather keep the use of overloading.
I suggest we keep it that way for now and change it later if it somehow becomes a problem.
We can also see what the others think about it

File: app/src/main/java/com/studhub/app/domain/usecase/listing/GetListings.kt (hunk 1 above)
Comment (edouardmichelin) [reply app/src/main/java/com/studhub/app/domain/usecase/listing/GetListings.kt:14]: Good catch

File: app/src/test/java/com/studhub/app/ListingUseCaseTest.kt [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        }
+    }
+
+    @After
+    fun clearDB() {
+        listingDB.clear()
+    }
+
+    @Test
+    fun `Get Listings Use Case retrieves all entries of the given repository`() = runBlocking {
```
Comment (raynasr) [inline app/src/test/java/com/studhub/app/ListingUseCaseTest.kt:83]: More efficient use of runBlocking {}

// here setup 

val getListings = GetListings(repository)

    val productId = Random.nextLong()
    val productName = Random.nextLong().toString()
    val product1 = Listing(id = 1L, name = "Product 1", seller = User())
    val product2 = Listing(id = productId, name = productName, seller = User())
    listingDB[product1.id] = product1
    listingDB[product2.id] = product2

    // runBlocking only needed here
    val result = runBlocking {
        getListings().toList()
    }

    // now verify
    assertThat(result.size).isEqualTo(2)
    assertThat(result[0]).isEqualTo(ApiResponse.Loading)
    assertThat(result[1]).isInstanceOf(ApiResponse.Success::class.java)
    val successResult = result[1] as ApiResponse.Success<List<Listing>>
    assertThat(successResult.data).containsExactlyInAnyOrder(product1, product2)

File: app/src/test/java/com/studhub/app/ListingUseCaseTest.kt [hunk 4]
```diff
@@ -0,0 +1,228 @@
+package com.studhub.app
```
Comment (raynasr) [inline app/src/test/java/com/studhub/app/ListingUseCaseTest.kt:1]: Add edge case for empty Listing | what happens if we attempt to remove/update an empty listing

Comment (raynasr) [inline app/src/test/java/com/studhub/app/ListingUseCaseTest.kt:1]: Seperate setup code

"           val productId = Random.nextLong()
            val productName = Random.nextLong().toString()
            val listing = Listing(id = productId, name = productName, seller = User()) "

Comment (raynasr) [inline app/src/test/java/com/studhub/app/ListingUseCaseTest.kt:1]: Use runTest{} instead of runBlocking{} for test blocks with no delays (runTest skips delays). Incapsulate the whole test with runTest{} if you choose run test. For runBlocking, only incapsulate the parts which require execution.

File: app/src/main/java/com/studhub/app/domain/usecase/listing/RemoveListing.kt (hunk 2 above)
Comment (raynasr) [reply app/src/main/java/com/studhub/app/domain/usecase/listing/RemoveListing.kt:30]: Alright then. We could keep it until the next version if a problem arises unless someone else finds a problem with it now.

File: app/src/test/java/com/studhub/app/UserUseCaseTest.kt [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                    Assert.assertEquals(userName, user.userName)
+                }
+                is ApiResponse.Failure -> Assert.fail("Request failure")
+                is ApiResponse.Loading -> {}
+            }
+        }
+    }
+
+    @Test
+    fun `Update User Use Case retrieves the correct entry of the given repository`() = runBlocking {
```
Comment (raynasr) [inline app/src/test/java/com/studhub/app/UserUseCaseTest.kt:113]: The main idea being checked is if the user is updated. The repository is a detail of the test. Maybe two separate tests or just change the name to " Update User Use Case successfully updates existing user in repository"

File: app/src/test/java/com/studhub/app/UserUseCaseTest.kt [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+                if (userDB.containsKey(userId)) {
+                    val newListing = updatedUser.copy(id = userId)
+                    userDB[userId] = newListing
+                    emit(ApiResponse.Success(userDB.getValue(userId)))
+                } else
+                    emit(ApiResponse.Failure("No entry for this key"))
+            }
+        }
+
+        override suspend fun removeUser(userId: Long): Flow<ApiResponse<Boolean>> {
```
Comment (raynasr) [inline app/src/test/java/com/studhub/app/UserUseCaseTest.kt:57]: Shouldn't removeUser remove a user from userDB instead of returning Success at every call?

**Review by raynasr — CHANGES_REQUESTED** [review]
Optimize runBlocking{} use, check other comments for suggestions

**Review by tepoorte — APPROVED** [review]
The structure for the domain is articulate and understandable, still you can add some more specific description for the listing Classes from the repository.

File: app/src/test/java/com/studhub/app/ListingUseCaseTest.kt (hunk 4 above)
Comment (edouardmichelin) [reply app/src/test/java/com/studhub/app/ListingUseCaseTest.kt:1]: Here the tests are only to test a correct linking of the repository with the use cases, not test actual correctness of the repository implementation (which is implemented locally).

File: app/src/test/java/com/studhub/app/UserUseCaseTest.kt (hunk 6 above)
Comment (edouardmichelin) [reply app/src/test/java/com/studhub/app/UserUseCaseTest.kt:57]: This is just a stub implementation as, as you correctly mentioned, removeUserUseCase is not used/needed

File: app/src/test/java/com/studhub/app/UserUseCaseTest.kt (hunk 5 above)
Comment (edouardmichelin) [reply app/src/test/java/com/studhub/app/UserUseCaseTest.kt:113]: As already explained, here the purpose is only to test good linking of the use case with the repository, not to test any implementation correctness

**Review by raynasr — APPROVED** [review]
Good for me
